engine/physics/src/physics/physics_2d.cpp

```diff
@@ -1234,56 +1234,84 @@ namespace dmPhysics
-			fixture = fixture->GetNext();
-		}
-	}
-
-	bool GetMaskBit2D(HCollisionObject2D collision_object, uint16_t groupbit) {
-		b2Fixture* fixture = ((b2Body*)collision_object)->GetFixtureList();
-		if (fixture) {
-			if (fixture->GetType() != b2Shape::e_grid) {
-				b2Filter filter = fixture->GetFilterData(0);
-				return !!(filter.maskBits & groupbit);
-			}
-		}
-		return false;
-	}
+        b2Fixture* fixture = ((b2Body*)collision_object)->GetFixtureList();
+        while (fixture) {
+            // do sth with the fixture
+            if (fixture->GetType() != b2Shape::e_grid) {
+                b2Filter filter = fixture->GetFilterData(0); // all non-grid shapes have only one filter item indexed at position 0
+                filter.categoryBits = groupbit;
+                fixture->SetFilterData(filter, 0);
+            }
+            fixture = fixture->GetNext();   // NOTE: No guard condition in loop. Assumes proper state of Box2D fixture list.
+        }
+    }
+
+    uint16_t GetGroup2D(HCollisionObject2D collision_object) {
+        b2Fixture* fixture = ((b2Body*)collision_object)->GetFixtureList();
+        if (fixture) {
+            if (fixture->GetType() != b2Shape::e_grid) {
+                b2Filter filter = fixture->GetFilterData(0);
+                return filter.categoryBits;
+            }
+        }
+        return 0;
+    }
+
+    // updates a specific group bit of a collision object's current mask
+    void SetMaskBit2D(HCollisionObject2D collision_object, uint16_t groupbit, bool boolvalue) {
+        b2Fixture* fixture = ((b2Body*)collision_object)->GetFixtureList();
+        while (fixture) {
+            // do sth with the fixture
+            if (fixture->GetType() != b2Shape::e_grid) {
+                b2Filter filter = fixture->GetFilterData(0); // all non-grid shapes have only one filter item indexed at position 0
+                if (boolvalue)
+                    filter.maskBits |= groupbit;
+                else
+                    filter.maskBits &= ~groupbit;
+                fixture->SetFilterData(filter, 0);
+            }
+            fixture = fixture->GetNext();
+        }
+    }
+
+    bool UpdateMass2D(HCollisionObject2D collision_object, float mass) {
+        b2Body* body = (b2Body*)collision_object;
+        if (body->GetType() != b2_dynamicBody) {
+            return false;
+        }
+        b2Fixture* fixture = body->GetFixtureList();
+        float total_area = 0.0f;
+        while (fixture) {
+            b2MassData mass_data;
+            fixture->GetShape()->ComputeMass(&mass_data, 1.0f);
+            // Since density is 1.0, massData.mass represents the area.
+            total_area += mass_data.mass;
+            fixture = fixture->GetNext();
+
+        }
+        fixture = body->GetFixtureList();
+        if (total_area <= 0.0f) {
+            return false;
+        }
+        float new_density = mass / total_area;
+        while (fixture) {
+            fixture->SetDensity(new_density);
```

Comment:
Is density the same for all fixtures _before_ updating the mass?
Because, now they are.
Also, I can totally see us changing this in the future, where each shape should have it's own mass.

Reply:
There is no such option to set mass per fixture. The only thing we can add in the future is to set density manually per fixture and then calculate mass based on it (or vice-versa). But it looks like some advanced level of setup. When I checked the other box2d tools/engines - nobody setup it like this.

Before - yes, they are set equal during creating, but with wrong collision object mass per shape (we will fix later).

Reply:
No, density is different for each shape in current implementation. It was more like the same mass for each piece with different density. But it is still one collision object, why it should have different density unless you manually decided to have it different (which isn't possible right now)?

I mean, you have a collision object with 2 shapes and mass 2kg. The first shape is square 100x100 and the second 1x1. Why should they have the same mass if the first one has 10k bigger area?